hppc/src/main/templates/com/carrotsearch/hppc/SortedIterationKTypeVTypeHashMap.java

```diff
@@ -0,0 +1,527 @@
+/*! #set($TemplateOptions.ignored = ($TemplateOptions.isKTypeAnyOf("DOUBLE", "FLOAT", "BYTE"))) !*/
+package com.carrotsearch.hppc;
+
+import com.carrotsearch.hppc.comparators.*;
+import com.carrotsearch.hppc.cursors.*;
+import com.carrotsearch.hppc.predicates.*;
+import com.carrotsearch.hppc.procedures.*;
+import com.carrotsearch.hppc.sorting.IndirectSort;
+
+/*! #if ($TemplateOptions.KTypeGeneric) !*/
+import java.util.Comparator;
+/*! #end !*/
+import java.util.Iterator;
+import java.util.function.IntBinaryOperator;
+
+/**
+ * Read-only view with sorted iteration order on a delegate {@link KTypeVTypeHashMap}.
+ *
+ * <p>In its constructor, this view creates its own iteration order array and sorts it,
+ * which is in O(n.log(n)) of the size of the delegate map. Afterward, calls to any method
+ * have the same performance as the delegate map.
+ *
+ * <p>This view does not support any modifying method. In addition, the delegate map must
+ * not be modified while the view is used, otherwise the iteration is undefined.
+ */
+/*! #if ($TemplateOptions.anyGeneric) @SuppressWarnings("unchecked") #end !*/
+/*! ${TemplateOptions.generatedAnnotation} !*/
+public class SortedIterationKTypeVTypeHashMap<KType, VType>
+  implements KTypeVTypeMap<KType, VType>
+{
+  public final KTypeVTypeHashMap<KType, VType> delegate;
+  public final int[] iterationOrder;
+
+  /**
+   * Creates a read-only view with sorted iteration order on the given delegate map.
+   * The ordering is in natural order of the keys.
+   */
+  public SortedIterationKTypeVTypeHashMap(KTypeVTypeHashMap<KType, VType> delegate) {
+    this.delegate = delegate;
+    this.iterationOrder = sortIterationOrder(createIterationOrder());
+  }
+
+  /*! #if ($TemplateOptions.KTypeGeneric) !*/
+  /**
+   * Creates a read-only view with sorted iteration order on the given delegate map.
+   * The ordering is based on the provided comparator on the keys.
+   */
+  public SortedIterationKTypeVTypeHashMap(KTypeVTypeHashMap<KType, VType> delegate,
+                                          Comparator<KType> comparator) {
+    this.delegate = delegate;
+    int[] iterationOrder = createIterationOrder();
+    this.iterationOrder = sortIterationOrder(iterationOrder, comparator);
+  }
+  /*! #end !*/
+
+  /**
+   * Creates a read-only view with sorted iteration order on the given delegate map.
+   * The ordering is based on the provided comparator on keys and values.
+   */
+  public SortedIterationKTypeVTypeHashMap(KTypeVTypeHashMap<KType, VType> delegate,
+                                          KTypeVTypeComparator<KType, VType> comparator) {
+    this.delegate = delegate;
+    int[] iterationOrder = createIterationOrder();
+    this.iterationOrder = sortIterationOrder(iterationOrder, comparator);
+  }
+
+  private int[] createIterationOrder() {
```

Comment:
maybe createEntryIndexes or something? The iteration order int array is on top of this so maybe worth distinguishing the two.